### PR TITLE
HEOMSolver return states and progress bar

### PR DESCRIPTION
**Pull request by christian512**
**Description**
While working on the [heom-tls notebook] I discovered two wrong behaviors of the `HEOMSolver`. 
1. Running `hsolver.run()` does not return the states if `e_ops=None`, but the documentation says that it should do so. I added the missing if condition to `HEOMSolver`. 
2. If passing a progress bar like `qutip.ui.progressbar.TextProgressBar()` to `HEOMSolver` it results in an error, because internally it only checks for `None` or `True`. The documentation says it should also work if I pass an instance of `BaseProgressBar` (or a subclass). I added the checks and a test in the testfile for this.

**Related issues or PRs**
None

**Changelog**
Fix HEOMSolver state return and progress bar

### CODE REVIEW

File: qutip/nonmarkov/bofin_solvers.py [hunk 1]
```diff
@@ -455,10 +455,16 @@ def __init__(
             for k in range(self._n_exponents)
         ]
 
+        self.progress_bar = progress_bar
         if progress_bar is None:
             self.progress_bar = BaseProgressBar()
         if progress_bar is True:
             self.progress_bar = TextProgressBar()
+        if not isinstance(self.progress_bar, BaseProgressBar):
```
Comment (hodgestar) [inline qutip/nonmarkov/bofin_solvers.py:463]: Thanks for sorting this out. I would recommend we change it a little more to be:
```
if progress_bar is None:
    ...
elif progress_bar is True:
    ...
elif isinstance(self.progress_bar, BaseProgressBar):
    ...
else:
    raise TypeError(...)
```
just to make it easy to see that the logic is correct.

**Review by hodgestar — CHANGES_REQUESTED** [review]
Thank you for fixing these and for adding the test. I requested one small further improvement to the progress bar setting, but otherwise it looks good.

File: qutip/nonmarkov/bofin_solvers.py (hunk 1 above)
Comment (christian512) [reply qutip/nonmarkov/bofin_solvers.py]: Thanks for the feedback :) I changed the if-statement to this cleaner version.

File: qutip/tests/test_bofin_solvers.py [hunk 2]
```diff
@@ -528,6 +528,9 @@ def test_create_progress_bar(self):
         hsolver = HEOMSolver(H, bath, 2, progress_bar=True)
         assert isinstance(hsolver.progress_bar, TextProgressBar)
 
+        hsolver = HEOMSolver(H, bath, 2, progress_bar=TextProgressBar())
+        assert isinstance(hsolver.progress_bar, BaseProgressBar)
+
```
Comment (hodgestar) [inline qutip/tests/test_bofin_solvers.py:533]: Could we also add a test for case where TypeError is raised?

File: qutip/nonmarkov/bofin_solvers.py [hunk 3]
```diff
@@ -457,8 +457,15 @@ def __init__(
...
         if progress_bar is None:
             self.progress_bar = BaseProgressBar()
-        if progress_bar is True:
+        elif progress_bar is True:
             self.progress_bar = TextProgressBar()
+        elif isinstance(progress_bar, BaseProgressBar):
+            self.progress_bar = progress_bar
+        else:
+            raise TypeError(
+                "ProgressBar (progress_bar) is not and instance of"
```
Comment (AGaliciaMartinez) [inline qutip/nonmarkov/bofin_solvers.py:466]:
```suggestion
                "ProgressBar (progress_bar) is not and instance of "
```

File: qutip/nonmarkov/bofin_solvers.py [hunk 4]
```diff
@@ -457,8 +457,15 @@ def __init__(
...
         if progress_bar is None:
             self.progress_bar = BaseProgressBar()
-        if progress_bar is True:
+        elif progress_bar is True:
             self.progress_bar = TextProgressBar()
+        elif isinstance(progress_bar, BaseProgressBar):
+            self.progress_bar = progress_bar
+        else:
+            raise TypeError(
+                "ProgressBar (progress_bar) is not and instance of "
```
Comment (hodgestar) [inline qutip/nonmarkov/bofin_solvers.py:466]:
```suggestion
                "ProgressBar (progress_bar) is not an instance of "
```

File: qutip/nonmarkov/bofin_solvers.py [hunk 5]
```diff
@@ -457,8 +457,15 @@ def __init__(
...
+                "qutip.ui.BaseProgressBar"
```
Comment (hodgestar) [inline qutip/nonmarkov/bofin_solvers.py:467]:
```suggestion
                "progress_bar is not an instance of "
                "qutip.ui.BaseProgressBar"
```

**Review by hodgestar — APPROVED** [review]

**Pull request merged**